### PR TITLE
[MM-52846] Migrate "components/autocomplete_selector.jsx" and tests to Typescript

### PR DESCRIPTION
**Pull request by janlengyel**
#### Summary

PR migrates AutocompleteSelector and it's tests from Javascript to Typescript

#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/23414

#### Screenshots

None

#### Release Note

```release-note
None
```

### CODE REVIEW

File: webapp/channels/src/components/autocomplete_selector.test.tsx [hunk 1]
```diff
@@ -60,7 +60,7 @@ describe('components/widgets/settings/AutocompleteSelector', () => {
             />,
         );
 
-        wrapper.instance().onBlur();
+        (wrapper.instance() as AutocompleteSelector).onBlur();
```
Comment (larkox) [inline webapp/channels/src/components/autocomplete_selector.test.tsx:63]: You should be able to pass the type directly to the shallow, and avoid these castings:

```
const wrapper = shallow<AutocompleteSelector>(
```

File: webapp/channels/src/components/autocomplete_selector.tsx [hunk 2]
```diff
@@ -1,55 +1,68 @@
...
+    listComponent?: typeof SuggestionList | typeof ModalSuggestionList;
+    listPosition?: string;
+};
+
+type State = {
+    input: string;
+    focused?: boolean;
+};
+
+type Selected = {
```
Comment (larkox) [inline webapp/channels/src/components/autocomplete_selector.tsx:34]: It is a bit out of the scope, but I have mixed feelings with this.

"Selected" is defined all over the place.

In `action_menu` is defined as `Option | UserProfile | Channel`
In `select_binding` is defined as `Option`
In `dialog_element` is defined as the way it is defined here.

We should centralize this type somehow (probably in this file, since it should be used only when the autocomplete selector is imported), and use always the same.

I feel the correct type in this case is `Option | UserProfile | Channel`, but 0/5.

Thoughts @M-ZubairAhmed ?

File: webapp/channels/src/components/autocomplete_selector.tsx [hunk 3]
```diff
@@ -108,21 +127,15 @@ export default class AutocompleteSelector extends React.PureComponent {
         let labelContent;
         if (label) {
             labelContent = (
-                <label
-                    className={'control-label ' + labelClassName}
-                >
+                <label className={'control-label ' + labelClassName}>
```
Comment (larkox) [inline webapp/channels/src/components/autocomplete_selector.tsx:130]: This and the next change seem unrelated.

File: webapp/channels/src/components/autocomplete_selector.tsx [hunk 4]
```diff
@@ -96,7 +109,13 @@ export default class AutocompleteSelector extends React.PureComponent {
             disabled,
             listComponent,
             listPosition,
-        } = this.props;
```
Comment (larkox) [inline webapp/channels/src/components/autocomplete_selector.tsx:99]: There is no need for this in order to generate the default values. Instead of doing:
```
const {foo, faa} = {foo = 1, ...props}
```
You can do:
```
const {foo = 1, faa} = props;
```

File: webapp/channels/src/components/autocomplete_selector.tsx [hunk 5]
```diff
@@ -96,7 +109,13 @@ export default class AutocompleteSelector extends React.PureComponent {
...
+        } = {
+            labelClassName: '',
```
Comment (larkox) [inline webapp/channels/src/components/autocomplete_selector.tsx:113]: Why don't you assign the default value for `value` here too? Why do you make it required when it was not before?

File: webapp/channels/src/components/autocomplete_selector.tsx [hunk 6]
```diff
@@ -1,55 +1,68 @@
...
+
+type Selected = {
+    id: string;
+    username: string;
+    display_name: string;
+    value: string;
+    text: string;
+}
 
-    static defaultProps = {
```
Comment (larkox) [inline webapp/channels/src/components/autocomplete_selector.tsx:27]: I have mixed feelings about removing the defualtProps. You made the behavior the same in the render, but I wonder if we ever come back to this component to add more functionality, if we will miss the defaults in any other method. 0/5

File: webapp/channels/src/components/post_view/embedded_bindings/select_binding/select_binding.tsx [hunk 7]
```diff
@@ -171,7 +171,7 @@ class SelectBinding extends React.PureComponent<Props, State> {
                         onSelected={this.handleSelected}
                         placeholder={label}
                         inputClassName='post-attachment-dropdown'
-                        value={this.state.selected?.text}
+                        value={this.state.selected?.text ?? ''}
```
Comment (larkox) [inline webapp/channels/src/components/post_view/embedded_bindings/select_binding/select_binding.tsx:174]: If we make `value` optional, this change will not be needed.

**Review by larkox — CHANGES_REQUESTED** [review]
Thank you for your contribution! A few concerns here and there.

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 2 above)
Comment (janlengyel) [reply webapp/channels/src/components/autocomplete_selector.tsx:34]: I also believe the correct type is `Option | UserProfile | Channel`. I also used it at first, but it required changes in the files you mentioned, so I went with this solution which didn't require changes.

File: webapp/channels/src/components/autocomplete_selector.test.tsx (hunk 1 above)
Comment (janlengyel) [reply webapp/channels/src/components/autocomplete_selector.test.tsx]: Thank you, I missed this.

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 3 above)
Comment (janlengyel) [reply webapp/channels/src/components/autocomplete_selector.tsx]: Changes reverted

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 6 above)
Comment (janlengyel) [reply webapp/channels/src/components/autocomplete_selector.tsx:27]: `defaultProps` are added back. I also removed the defaults in render.

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 5 above)
Comment (janlengyel) [reply webapp/channels/src/components/autocomplete_selector.tsx]: Made value optional, I set is as required because in `propTypes` it was `PropTypes.string.isRequired`.

File: webapp/channels/src/components/post_view/embedded_bindings/select_binding/select_binding.tsx (hunk 7 above)
Comment (janlengyel) [reply webapp/channels/src/components/post_view/embedded_bindings/select_binding/select_binding.tsx]: Value is now optional, change reverted.

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 4 above)
Comment (janlengyel) [reply webapp/channels/src/components/autocomplete_selector.tsx]: Thank you, I didn't know it works like that. Because I added the `defaultProps` back, the defaults are removed.

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 2 above)
Comment (janlengyel) [reply webapp/channels/src/components/autocomplete_selector.tsx:34]: I have the change ready, I can push it if needed.

File: webapp/channels/src/components/autocomplete_selector.tsx [hunk 8]
```diff
@@ -1,29 +1,49 @@
...
-        listComponent: PropTypes.elementType,
-        listPosition: PropTypes.string,
-    };
+import type ModalSuggestionList from './suggestion/modal_suggestion_list';
+import type Provider from './suggestion/provider';
+
+type Props = {
+    id?: string;
+    providers: Provider[];
+    value?: string;
```
Comment (larkox) [inline webapp/channels/src/components/autocomplete_selector.tsx:15]: In theory, if defaultProps is defined, the props don't need to be optional. Typescript is smart enough to consider them optional on the outside, and "defined" in the inside.

So... you can set back all props in the default props to required.

(If this doesn't work as I said, let me know, and we can look into it)

File: webapp/channels/src/components/autocomplete_selector.tsx [hunk 9]
```diff
@@ -101,7 +123,7 @@ export default class AutocompleteSelector extends React.PureComponent {
         const {focused} = this.state;
         let {input} = this.state;
 
-        if (!focused) {
+        if (!focused && value !== undefined) {
```
Comment (larkox) [inline webapp/channels/src/components/autocomplete_selector.tsx:126]: `value` is going to be always different to undefined, because it is set with the defaultProps. Hopefully updating the props as I said in my last comment will make this change unneeded.

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 5 above)
Comment (larkox) [reply webapp/channels/src/components/autocomplete_selector.tsx]: Oh, I see. Probably the prop types were not well defined :sweat_smile: Sorry about that

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 8 above)
Comment (janlengyel) [reply webapp/channels/src/components/autocomplete_selector.tsx]: Yes, it is working 🙂. I usually work with Angular and it's a while since I last used React, so I was not aware of that. Today I learned 😄.

**Review by larkox — APPROVED** [review]
Apart of the open discussion about `Selected`, looks good to me.

Thank you for the contribution!

File: webapp/channels/src/components/autocomplete_selector.tsx (hunk 2 above)
Comment (hmhealey) [reply webapp/channels/src/components/autocomplete_selector.tsx:34]: If the change doesn't cause other typing issues, I'd be good with that. Ideally, I think this component would have some parameterized type like `AutocompleteSelector<UserProfile>`, but React seems to have trouble with those, so using an option type to avoid repeating ourselves sounds good to me

**Review by hmhealey — CHANGES_REQUESTED** [review]

**Review by larkox — APPROVED** [review]
LGTM!

**Review by hmhealey — APPROVED** [review]

**Comment by hmhealey** [comment]
Thanks for helping with the migration. This looks good to me!

**Review by yasserfaraazkhan — APPROVED** [review]

**Comment by larkox** [comment]
Thank you @janlengyel for the contribution!
